pkg/utils/commands.go

```diff
@@ -0,0 +1,224 @@
+			return fmt.Errorf("ApplyPostCommands:: rule type [%s] not supportted", rule.Type)
+		}
+	}
+	for _, command := range commands {
+		err := c.ExecutePodCommand(rule.Name, []string{"sh", "-c", command})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ExecutePodCommand ...
+func (c *PodCommands) ExecutePodCommand(hookName string, command []string) error {
+	if err := c.validate(); err != nil {
+		return err
+	}
+	if len(command) == 0 {
+		return fmt.Errorf("ExecutePodCommand:: command is required")
+	}
+	pod, err := c.getContainers()
+	if err != nil {
+		return err
+	}
+	if c.container == "" {
+		c.setDefaultHookContainer(pod)
+	}
+	err = c.ensureContainerExists(pod)
+	if err != nil {
+		return err
+	}
+
+	hookLog := log.WithFields(
+		log.Fields{
+			"hookName":      hookName,
+			"hookContainer": c.container,
+			"hookCommand":   command,
+			"hookOnError":   "Fail",
+			"hookTimeout":   defaultTimeout,
+		},
+	)
+	hookLog.Info("ExecutePodCommand:: running exec hook")
+
+	req := c.client.CoreV1().RESTClient().Post().Resource("pods").Name(c.name).
+		Namespace(c.namespace).SubResource("exec")
+
+	req.VersionedParams(
+		&corev1api.PodExecOptions{
+			Command: command,
+			Stdout:  true,
+			Stderr:  true,
+		}, scheme.ParameterCodec,
+	)
+	exec, err := c.streamExecutorFactory.NewSPDYExecutor(c.restConfig, "POST", req.URL())
+	if err != nil {
+		return err
+	}
+	var stdout, stderr bytes.Buffer
+	errChan := make(chan error)
+	go func() {
+		err = exec.Stream(remotecommand.StreamOptions{
+			Stdout: &stdout,
+			Stderr: &stderr,
+		})
+		errChan <- err
+	}()
+
+	var timeoutCh <-chan time.Time
+	timer := time.NewTimer(defaultTimeout)
+	defer timer.Stop()
+	timeoutCh = timer.C
+
+	select {
+	case err = <-errChan:
+	case <-timeoutCh:
+		return fmt.Errorf("ExecutePodCommand:: timeout after: %v", defaultTimeout)
+	}
+	log.Infof("ExecutePodCommand:: stdout: %v", stdout.String())
+	log.Infof("ExecutePodCommand:: stderr: %v", stderr.String())
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *PodCommands) getContainers() (*corev1api.Pod, error) {
+	pod, err := c.client.CoreV1().Pods(c.namespace).Get(context.Background(), c.name, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("getContainers:: failed to get pods: %v", err)
+	}
+	return pod, nil
+}
+
+func (c *PodCommands) ensureContainerExists(pod *corev1api.Pod) error {
+	if len(pod.Spec.Containers) > 0 {
+		for _, container := range pod.Spec.Containers {
+			if container.Name == c.container {
+				return nil
+			}
+		}
+	} else {
+		return fmt.Errorf("ensureContainerExists:: pod containers count is zero: %v", pod.Spec.Containers)
+	}
+	return fmt.Errorf("ensureContainerExists:: container name [%s] not in [%s] pod", c.container, c.name)
+}
+
+func (c *PodCommands) setDefaultHookContainer(pod *corev1api.Pod) error {
+	if len(pod.Spec.Containers) < 1 {
+		return fmt.Errorf("need at least 1 container")
+	}
+
+	c.container = pod.Spec.Containers[0].Name
```

Comment:
Containers[0] does not means the main database container for a multiple containers application, a sidecar and other container. container name should be specified in spec of rule or snapshot spec explicitly.